From c65dce4c6d8d54e47dce79a56e29e2223a2354e6 Mon Sep 17 00:00:00 2001
From: Vikram Gaur <vikramgaur@google.com>
Date: Fri, 17 May 2024 23:34:58 +0000
Subject: [PATCH] RESTRICT AUTOMERGE: Allow System user to call getRegistration
 API.

Bug: 336976105
Test: Verified on CF that only system server is able to bind.
Ignore-AOSP-First: Backporting fix
Flag: EXEMPT Mainline module change.
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:d9c9db03fae63ac16d76cd68450e78c4a9285104)
Merged-In: I6ee570b9c432ece71442cb989dd37cf5df35dc55
Change-Id: I6ee570b9c432ece71442cb989dd37cf5df35dc55
---
 app/AndroidManifest.xml                                |  1 +
 .../rkpdapp/service/RemoteProvisioningService.java     | 10 ++++++++++
 2 files changed, 11 insertions(+)

diff --git a/app/AndroidManifest.xml b/app/AndroidManifest.xml
index 79a1783..b8ab08b 100644
--- a/app/AndroidManifest.xml
+++ b/app/AndroidManifest.xml
@@ -30,6 +30,7 @@
             android:exported="false">
         </service>
         <service android:name=".service.RemoteProvisioningService"
+            android:permission="android.permission.BIND_RKP_SERVICE"
             android:exported="true">
             <intent-filter>
                 <action android:name="com.android.rkpdapp.IRemoteProvisioning"/>
diff --git a/app/src/com/android/rkpdapp/service/RemoteProvisioningService.java b/app/src/com/android/rkpdapp/service/RemoteProvisioningService.java
index 544d59d..ae96686 100644
--- a/app/src/com/android/rkpdapp/service/RemoteProvisioningService.java
+++ b/app/src/com/android/rkpdapp/service/RemoteProvisioningService.java
@@ -19,7 +19,9 @@ package com.android.rkpdapp.service;
 import android.app.Service;
 import android.content.Context;
 import android.content.Intent;
+import android.os.Binder;
 import android.os.IBinder;
+import android.os.Process;
 import android.os.RemoteException;
 import android.util.Log;
 
@@ -64,6 +66,14 @@ public class RemoteProvisioningService extends Service {
                     return;
                 }
 
+                // Check that only system process and self calls binding.
+                if (Binder.getCallingUid() != Process.SYSTEM_UID
+                        && Binder.getCallingUid() != Process.myUid()) {
+                    callback.onError(
+                            "Only system server and self are allowed to call RKP service.");
+                    return;
+                }
+
                 SystemInterface systemInterface;
                 try {
                     systemInterface = ServiceManagerInterface.getInstance(irpcName);
-- 
2.46.0.rc2.264.g509ed76dc8-goog

